rendering-event-loop.md

```diff
@@ -0,0 +1,94 @@
+
+### The browser context (aka rendering) event loop
+
+There are several event loops, but the most important is the
+[browsing context](https://html.spec.whatwg.org/multipage/browsers.html#browsing-context) event loop. This event loop is special because it is where tasks related to
+* document-loaded scripts (as opposed to worker scripts),
+* DOM,
+* input event processing
+* rendering updates
+
+all occur.
+
+As it relates to rendering, the spec currently defines image decoding, video,
+and HTMLCanvasElement.{toBlob,convertToBlob} to be in parallel.
+
+However, any task which affects script-observable data (i.e. specific to a
+particular script realm, global or evironment settings object) [must not be in parallel](https://html.spec.whatwg.org/#event-loop-for-spec-authors) to the task queue on which such a script runs. This means that:
+* Script can not be run in parallel with script from the same realm
+* DOM (which is script-observable) cannot be mutated in parallel with script
+
+## Rendering: step 7 of the browsing context event loop
+
+Step 7 of the [event loop processing model](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model), is the steps to
+get ready and "update the rendering". This is spec'ed to potentially happen
+at the end of processing every task in a browsing context event loop.
+However, it is also specified that these steps need not happen after every
+task, and may be throttled to a specified framerate, or for other reasons.
+
+The update-the-rendering steps in the HTML specification are:
+
+1. Resize
+2. Scroll
+3. Media queries
+4. Update animations, fire animation events
+5. Fire fullscreen events
+6. Run animation frame callbacks
+7. Update intersection observations
+8. "Update the rendering" (draw things to the screen)
+
+The spec has no concept of style or layout computation. It's assumed to be
+synchronously implied by DOM changes, even before step 1 above happens.
+
+
+# Rendering in parallel
+
+The steps of rendering (i.e. steps 1-8 from the previous section) must happen in series, and if rendering is updated they must all be run. In addition, the spec does not say that these steps are in parallel with other tasks. However, as discussed above, the only thing preventing tasks or steps from running in parallel in principle is script observability, which is in turn due to the inherently single-threaded nature of JavaScript.
+
+Note also that *script observability* (a script program can see the difference) is different than *user observability* (the person viewing the web site can see something happening). Thus if we are willing to let the final rendering to the screen be temporarily out of sync (i.e. what the user can see is different than what the script last saw), there is an opportunity for parallelism. (Note that the HTML spec makes no reference to this possibility, though it does not preclude it either.)
```

Comment:
The fact that the consistency between script and user observability isn't specced is a bit surprising to me

Reply:
I think it's not spec'ed on purpose, to give flexibility to user agents.